rubicon/objc/collections.py

```diff
@@ -1,10 +1,159 @@
-from .types import NSUInteger, NSNotFound, NSRange
+import operator
+
+from .types import NSUInteger, NSNotFound, NSRange, unichar
 from .runtime import (
-    NSArray, NSDictionary, NSMutableArray, NSMutableDictionary, ObjCInstance, for_objcclass, ns_from_py, objc_id,
-    send_message
+    NSArray, NSDictionary, NSMutableArray, NSMutableDictionary, NSString, ObjCInstance, for_objcclass, ns_from_py,
+    objc_id, py_from_ns, send_message,
 )
 
 
+# Some useful NSStringCompareOptions.
+NSLiteralSearch = 2
+NSBackwardsSearch = 4
+
+
+@for_objcclass(NSString)
+class ObjCStrInstance(ObjCInstance):
+    """Provides Pythonic operations on NSString objects that mimic those of Python's str.
+
+    Note that str objects consist of Unicode code points, whereas NSString objects consist of UTF-16 code units.
+    These are not equivalent for code points greater than U+FFFF. For performance and simplicity, ObjCStrInstance
+    objects behave as sequences of UTF-16 code units, like NSString. (Individual UTF-16 code units are represented as
+    Python str objects of length 1.) If you need to access or iterate over code points instead of UTF-16 code units,
+    use str(nsstring) to convert the NSString to a Python str first.
+    """
+
+    def __str__(self):
+        return self.UTF8String.decode('utf-8')
+
+    def __eq__(self, other):
+        if isinstance(other, str):
+            return self.isEqualToString(ns_from_py(other))
+        elif isinstance(other, NSString):
+            return self.isEqualToString(other)
+        else:
+            return super().__eq__(other)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    # Note: We cannot define a __hash__ for NSString objects; doing so would violate the Python convention that
+    # mutable objects should not be hashable. Although we could disallow hashing for NSMutableString objects, this
+    # would make some immutable strings unhashable as well, because immutable strings can have a runtime class that
+    # is a subclass of NSMutableString. This is not just a theoretical possibility - for example, on OS X 10.11,
+    # isinstance(NSString.string(), NSMutableString) is true.
```

Comment:
Ok - We should probably document that limitation (and the 'cast to Python' workaround) somewhere to avoid confusion.

Reply:
We need to document more things in general 😃 I'm actually not quite sure what the best place for this information would be. I suppose I could put it in the type conversion howto, since `ObjCDictInstance` and `ObjCListInstance` are explained there too, but I feel like this sort of thing would be better suited for the reference documentation.

I don't think the behavior is that strange either, since normal `ObjCInstance` and all other subclasses are also unhashable. On the other hand, I can see why one might expect it to work, since strings work fine as dictionary keys in pure Python and pure ObjC.

Reply:
OK, done. Part of the new "Strings" section in the docs is copy-pasted from the "Lists" section below (with minor modifications), but I think that's okay for now. I was planning to work on the docs next anyway, so when I do that I'll see if I can rewrite the "Strings", "Lists" and "Dictionaries" sections a bit so they don't repeat themselves as much.